### ISSUE TITLE
Forward references in saving

### ISSUE THREAD

**Issue by JacobHast** [qua-platform/quam#90]
If I have a component with an attribute using forward referencing,

e.g.
```
@quam_dataclass
class ComponentA(QuamComponent):
    x: int

@quam_dataclass
class ComponentB(QuamComponent):
     component_a = "ComponentA"
```

I get a ComponentA is not defined NameError when calling .save()

At the moment this is an issue for us, as we are using qubits and resonators which have resonator and qubit attributes respectively, and so to avoid circular imports we use forward referencing, which seems to work well, except that we can save our state.

**Comment by nulinspiratie** [comment]
Forward referencing is indeed an issue, partly because determining the correct class from the string is problematic.
Can I suggest to first try adding `if TYPE_CHECKING:` and see if that resolves the issue. So for example:

### resonator.py
```python
from typing import TYPE_CHECKING
from quam.core import QuamComponent, quam_dataclass

if TYPE_CHECKING:
    from quam.examples.superconducting_qubits.qubit import Qubit


@quam_dataclass
class Resonator(QuamComponent):
    qubit: Qubit
```

### qubit.py
```python
from typing import TYPE_CHECKING
from quam.core import QuamComponent, quam_dataclass

if TYPE_CHECKING:
    from quam.examples.superconducting_qubits.resonator import Resonator


@quam_dataclass
class Qubit(QuamComponent):
    resonator: Resonator
```

Let me know if this resolves your issue

**Comment by JacobHast** [comment]
This doesn't seem to work unfortunately, I get an error `name 'Resonator' is not defined` from the line `resonator: Resonator` when importing from qubit.py in your example.

**Comment by nulinspiratie** [comment]
@JacobHast you're right, I'm able to reproduce this. I was able to fix it by adding `from __future__ import annotations` to the top of the file, does that fix it for you as well?

**Comment by JacobHast** [comment]
Adding `from __future__ import annotations` works :)

**Comment by nulinspiratie** [comment]
Great! In that case I'll close this issue. I've looked a bit into adding support for forward references, and I have not come across a robust solution. If we don't come across a use case that requires this, I'm in favor of not implementing it

**Comment by JacobHast** [comment]
Hi Serwan, the solution causes issues when loading the state, it seems that when doing this it no longer correctly identifies the correct classes and so throws an error

**Comment by nulinspiratie** [comment]
Hmm odd, can you share the traceback so I know it matches?

**Comment by JacobHast** [comment]
Here's an example that produces an issue, I would think this is related to the loading issue I'm getting:

**qubit.py**
```
from __future__ import annotations
from typing import TYPE_CHECKING
from quam.core import QuamComponent, quam_dataclass


if TYPE_CHECKING:
    from resonator import Resonator


@quam_dataclass
class Qubit(QuamComponent):
    resonator: Resonator

```

**resonator.py**
```
from __future__ import annotations
from typing import TYPE_CHECKING
from quam.core import QuamComponent, quam_dataclass

if TYPE_CHECKING:
    from qubit import Qubit


@quam_dataclass
class Resonator(QuamComponent):
    qubit: Qubit

```

**main.py**
```
from qubit import Qubit
from resonator import Resonator
from quam import QuamRoot, quam_dataclass


@quam_dataclass
class Root(QuamRoot):
    qubit: Qubit
    resonator: Resonator


root = Root(
    qubit=Qubit(resonator=Resonator(qubit="#../../")), resonator="#./qubit/resonator"
)

# this fails with NameError: name 'Resonator' is not defined
root.to_dict()
```

**Comment by nulinspiratie** [comment]
I've been able to reproduce this, but honestly I haven't been able to solve it. I tried several methods but they all raise a similar issue. 
How critical of an issue is this? The only other solution I can think of is creating a custom function rather than use the typing package, but this is risky

**Comment by nulinspiratie** [comment]
Found a fix, let me know if this works for you
